obspy/signal/polarization.py

```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
+    modification of adapted analysis window estimated by estimating the
+    instantaneous frequency. It returns the azimuth, incidence, rectilinearity
+    planarity and ellipticity.
+
+    :param stream: ZNE sorted trace data
+    :type stream: :class:`~obspy.core.stream.Stream`
+    :param noise_thres: Variance of the noise sphere; data points are excluded
+        when falling within the sphere with radius sqrt(noise_thres),
+        Default = 0
+    :type noise_thres: float
+    :param fs: sampling rate
+    :type fs: float
+    :param flow: lower frequency for analysis
+    :type flow: float
+    :param fhigh: upper frequency limit for analysis
+    :type fhigh: float
+    :param spoint: array with traces` individual start times in sample
+    :type spoint: :class:`numpy.ndarray`
+    :param stime: start time of the analysis
+    :type stime: :class:`~obspy.core.utcdatetime.UTCDateTime`
+    :param etime: end time for the analysis
+    :type etime: :class:`~obspy.core.utcdatetime.UTCDateTime`
+    :returns: azimuth, incidence, rectilinearity planarity, and ellipticity
+    """
+    W = 3.0
+    stream.sort(reverse=True)
+    Z = stream[0].data.copy()
+    N = stream[1].data.copy()
+    E = stream[2].data.copy()
+
+    Zi = instantaneous_frequency(Z, fs)
+    Za = signal.hilbert(Z)
+    Ni = instantaneous_frequency(N, fs)
+    Na = signal.hilbert(N)
+    Ei = instantaneous_frequency(E, fs)
+    Ea = signal.hilbert(E)
+    res = []
+
+    offset = int(3 * fs / flow)
+    while True:
+        adapt = int(3. * W * fs / (Zi[offset] + Ni[offset] + Ei[offset]))
+        # in order to account for errors in the inst freq estimation
+        if adapt > int(3 * fs / flow):
+            adapt = int(3 * fs / flow)
+        if adapt < int(3 * fs / fhigh):
+            adapt = int(3 * fs / fhigh)
+        adapt /= 2
+        adapt = (2 * adapt) + 1
+        newstart = stime + offset / fs
+        if (newstart + (adapt / 2) / fs) > etime:
+            break
+
+        Zx = Za[int(spoint[2] + offset - adapt / 2):
+                int(spoint[2] + offset + adapt / 2)]
+        Nx = Na[int(spoint[1] + offset - adapt / 2):
+                int(spoint[1] + offset + adapt / 2)]
+        Ex = Ea[int(spoint[0] + offset - adapt / 2):
+                int(spoint[0] + offset + adapt / 2)]
+        Zx -= Zx.mean()
+        Nx -= Nx.mean()
+        Ex -= Ex.mean()
+
+        covmat = np.zeros([3, 3], dtype=np.complex128)
+
+        covmat[0][0] = np.dot(Ex, Ex.conjugate())
+        covmat[0][1] = np.dot(Ex, Nx.conjugate())
+        covmat[1][0] = covmat[0][1].conjugate()
+        covmat[0][2] = np.dot(Ex, Zx.conjugate())
+        covmat[2][0] = covmat[0][2].conjugate()
+        covmat[1][1] = np.dot(Nx, Nx.conjugate())
+        covmat[1][2] = np.dot(Zx, Nx.conjugate())
+        covmat[2][1] = covmat[1][2].conjugate()
+        covmat[2][2] = np.dot(Zx, Zx.conjugate())
+
+        eigvec, eigenval, v = (np.linalg.svd(covmat))
```

Comment:
Except for the `ellip` calculation, from here down appears to be quite similar to `flinn`. Might be useful to put that in a common function.